protocols/messages.go

```diff
@@ -1,19 +1,98 @@
 package protocols
 
 import (
+	"bytes"
 	"encoding/json"
+	"sort"
 
+	"github.com/ethereum/go-ethereum/common"
 	"github.com/statechannels/go-nitro/channel/consensus_channel"
 	"github.com/statechannels/go-nitro/channel/state"
 	"github.com/statechannels/go-nitro/types"
 )
 
+const (
+	SignedStatePayload    PayloadType = "SignedStatePayload"
+	SignedProposalPayload PayloadType = "SignedProposalPayload"
+)
+
+type PayloadType string
+
 // Message is an object to be sent across the wire. It can contain a proposal and signed states, and is addressed to a counterparty.
 type Message struct {
-	To              types.Address
-	ObjectiveId     ObjectiveId
-	SignedStates    []state.SignedState
-	SignedProposals []consensus_channel.SignedProposal
+	To       types.Address
+	Payloads []MessagePayload
+}
+
+// MessagePayload is an objective id and EITHER a SignedState or SignedProposal.
+type MessagePayload struct {
+	ObjectiveId    ObjectiveId
+	SignedState    state.SignedState
+	SignedProposal consensus_channel.SignedProposal
+}
+
+// hasState returns true if the payload contains a signed state.
+func (p MessagePayload) hasState() bool {
+	stateHash, err := p.SignedState.State().Hash()
+	if err != nil {
+		return false
+	}
+
+	return bytes.Compare(stateHash.Bytes(), common.Hash{}.Bytes()) != 0
+}
+
+// hasProposal returns true if the payload contains a signed proposal.
+func (p MessagePayload) hasProposal() bool {
+	return p.SignedProposal.Proposal != consensus_channel.Proposal{}
+}
+
+// Type returns the type of the payload, either a SignedProposal or SignedState.
+func (p MessagePayload) Type() PayloadType {
+
+	if p.hasProposal() && !p.hasState() {
+		return SignedProposalPayload
+	} else if p.hasState() && !p.hasProposal() {
+		return SignedStatePayload
+	}
+	panic("payload has both state and proposal")
+}
+
+// ObjectivePayload is a struct that contains an objectiveId and EITHER.
+type ObjectivePayload[T PayloadValue] struct {
+	Payload     T
+	ObjectiveId ObjectiveId
+}
+
+// SignedStates returns a slice of signed states with their objectiveId that were contained in the message.
+// The states are sorted by channel id then turnNum.
```

Comment:
Sorting by turnNum seems to have a clear purpose (it is part of our ledger protocol). But how important is it to sort by channel id?

Reply:
I don't think sorting by `ChannelId` is absolutely necessary. However it does mean that the proposals/states get sorted deterministically which seems nice.

It also seems easier to parse a list of states mentally if they are sorted first by channel id then turn number.